[mkcal] Remove unused loaded flags from ExtendedStorage.

dcaliste opened the pull request:
@pvuorela , I forgot to kill a bit more lines in ExtendedStorage : all these loaded flags are now useless.

dcaliste commented:
Not related to this PR, but since we're at deleting unused code, what about:
- `ExtendedStorage::fooCount()`, is there any use cases for these (like in ExtendedClendar) ?
- `ExtendedStorage::checkAlarm()`, I don't know if there is a use case besides the one in the documentation string. I mean, for this case, I should be safer if the alarm can provide the instance identifier or the UID (like I think it can be done at the moment with timed).

pvuorela commented:
> ExtendedStorage::fooCount(), is there any use cases for these (like in ExtendedClendar) ?

Wouldn't think so. Same vagueness problem also with that.

> ExtendedStorage::checkAlarm()

That's a strange method. Nothing needs it, but if something would, it should be quite straightforward just to implement the load() + hasEnabledAlarms() in the app side. Let's delete.

Can use JB#60291 for these mkcal api removals btw.

dcaliste commented:
@pvuorela, I've added 2 commits removing the fooCount() API and the checkAlarm(), as discussed earlier.

I'm also pushing 3 commits on top. Give me your opinion : they try to reduce as much as possible the public API of ExtendedStorage:
- clearLoaded() is now hidden from the children classes, since its meaning is more like the close() routine. As mentioned in the commit, if necessary implementation classes should call setModified() to ensure that all side actions like resetting the calendar notebooks are also done.
- I'm treating loadRecurringIncidences() as an implementation detail now. As far as I know, sync plugins are not using it and the QML bindings have already been migrated some while ago to use loadRange() only.
- I've moved the generic load() methods to ExtendedStorage with a default implementation there. If necessary, implementations providing a better way to load events than the generic way can still overload these methods.

I've added a last commit that also do event UID sanitising in only one place.

pvuorela commented:
I'll check the latest additions soon, but now when checking the earlier ones and the current code, couple more candidates:

The ExtendedStorage::duplicateIncidences() is strange. Name halfway verb or getter, but then again the "duplicate" being also a vague concept. Per se the whole system doesn't allow full duplicates to exist. Seems now being summary and start time being the same. Application side could do this if there's need, in this form seems useless, and unused too.

ExtendedStorage header has DBOperation exposed and cpp uses that for doing notebook operations on storage. I can live with that, but then the enum is also used for other needs in the SqliteStorage. For starters could be simpler if different cases used different enum. But the main case seems to be SqliteStorage::Private::selectIncidences() where DBInsert (for example) doesn't insert anything, but decides what value is bound to the query (even though the cases look pretty similar). Think this could be clearer if instead of query string the parameter would be sql statement and the caller would be responsible for binding the values.

dcaliste commented:
> The ExtendedStorage::duplicateIncidences() is strange.

I'll remove it happily. I see the use case of it : do not load the full DB when you want to look for some incidences only and delegate the search part to the DB. The proposed search API in another PR can do this in a more generic way. And I agree with you that it's up to the client code then to look into the search results and behave accordingly.

> ExtendedStorage header has DBOperation exposed

I'll see what I can do and surely follow your advice to split the statement preparation into calling routines.

dcaliste commented:
So, I've removed duplicateIncidences() query routine. I've also remove DBOperation from ExtendedStorage (it's now only in SqliteFormat for the modify{Components,Calendar}() rotuine). To do so, I've followed your advice to move the statement preparation and binding of the query routines inside the callers, and to split the virtual modifyNotebook() into the three possible actions.

The later may look a bit verbose, but it may disappear in the near future. Indeed, after all this cleaning, it appears clear to me that the notebook routines are misplaced. They belong to ExtendedCalendar, with proper signaling of insertion, modification and deletion, like for the incidences. Then, the storage can be an observer of the notebooks in the ExtendedCalendar and behave accordingly. Because currently we have the calendar that store the incidences and the storage that store the notebooks. It's a strange separation. The calendar should be the memory representation that the client is poking into, while the storage should be a serialiser to the DB only. With that vision, the new insertNotebook(), modifyNotebook() and eraseNotebook() will disappear and the storage implementation will simply call SqliteStorage::Private::saveNotebook() with the proper DBOperation as an internal implementation detail. Exactly like in the ::save() routine for the incidences.

pvuorela commented on src/sqlitestorage.cpp:
```diff
@@ -1039,8 +900,30 @@ bool SqliteStorage::insertedIncidences(Incidence::List *list, const QDateTime &a
+        SL3_prepare_v2(d->mDatabase, query1, qsize1, &stmt1, nullptr);
+        secs = d->mFormat->toOriginTime(after);
+        SL3_bind_int64(stmt1, index, secs);
+        if (!notebookUid.isNull()) {
```
Was already like that, but think isEmpty() could be more correct than isNull. Same thing on many other places here.

dcaliste replied:
Sure, changed in all the query methods.

pvuorela reviewed (commented):
A few small comments.

pvuorela commented on src/extendedstorage.cpp:
```diff
@@ -479,7 +391,7 @@ bool ExtendedStorage::updateNotebook(const Notebook::Ptr &nb)
     }
 
-    if (!nb->isRunTimeOnly() && !modifyNotebook(nb, DBUpdate)) {
+    if (!nb->isRunTimeOnly() && !modifyNotebook(nb)) {
```
Maybe updateNotebook() as name could be slightly clearer since the operation still maps to DBUpdate on lower level.

dcaliste replied:
The problem is that `addNotebook(Notebook::Ptr&)`,  `updateNotebook(Notebook::Ptr&)` and  `deleteNotebook(Notebook::Ptr&)` are already part of the public API with another meaning...

pvuorela replied:
Ah, of course.

dcaliste commented on src/extendedstorage.cpp:
```diff
@@ -129,6 +113,16 @@ class mKCal::ExtendedStorage::Private
 #endif
 };
+
+bool ExtendedStorage::Private::close()
```
We cannot call the public `close()` since this public method is virtual and we don't want to actually close the database access as done in `SqliteStorage::close()`. Which imply that the name is badly chosen, I you said. So I'm keeping this routine and simply changing its name to `clear()`.

dcaliste commented on src/sqlitestorage.cpp:
```diff
@@ -138,14 +138,11 @@ class mKCal::SqliteStorage::Private
     bool mIsSaved;
 
     bool addIncidence(const Incidence::Ptr &incidence, const QString &notebookUid);
+    int loadRecurringIncidences();
```
It's to follow the `LoadIncidences()` method in the same private class which also returns an int, but use it as a bool everywhere with `(count >= 0)`.

So I can keep it like that, or change also `loadIncidences()` for a bool. What do you prefer ?

pvuorela replied:
I'm ok with anything, but if the return value is interested only in success, the bool might be slightly simpler? Wouldn't mind with internal methods not having consistent bool vs int on these either.

dcaliste replied:
Ok, I've changed back the return value to bool, only for the loadRecurringIncidences().

dcaliste commented on src/extendedstorage.cpp:
```diff
@@ -393,9 +307,7 @@ void ExtendedStorage::setModified(const QString &info)
-    clearLoaded();
-    d->mNotebooks.clear();
-    d->mDefaultNotebook.clear();
+    d->close();
```
I don't like much the `info` argument but it's part of the public observer API and I'm not keen to delete it yet.  Would need some propagation to nemo-qml-plugin-calendar. Maybe for a later API break commit, all of those of the current PR should not require any propagation of API changes.

What I can do though is to rename these vague names with the name of the observer signal and an `emit` verb. Like that `setModified()` would become `emitStorageModified()`. It would make the purpose of calling such method clear in the implementations.

dcaliste commented on src/sqlitestorage.cpp:
```diff
@@ -730,7 +690,7 @@ bool SqliteStorage::save(ExtendedStorage::DeleteAction deleteAction)
     Incidence::List deleted;
     if (!d->mIncidencesToDelete.isEmpty()) {
-        DBOperation dbop = DBNone;
+        DBOperation dbop;
```
Ah yes, I always forget that values of type enum can get values outside the enum elements.

I've changed for `dbop = deleteAction == Delete ? DBDelete : DBMarkDeleted;`

dcaliste commented:
Thanks, I've corrected the typo.
